### PR TITLE
Add support for finding Atom, Visual Studio, and VS Code

### PR DESCRIPTION
**Pull request by shimpossible**
[WIP]
Allow for additional external editors like the old GitHub Desktop.  Supports #1554 #2009

### CODE REVIEW

File: app/src/lib/dispatcher/app-shell.ts [hunk 1]
```diff
@@ -1,15 +1,56 @@
...
+  result.push({
+    name: 'External Editor',
+  })
+  if (__DARWIN__) {
+    const osx = require('shell-osx')
+    if (osx.isAtomInstalled()) {
+    result.push({
+      name: 'Atom',
+    })
+    }
```
Comment (j-f1) [inline app/src/lib/dispatcher/app-shell.ts]: Can you check if VS Code is installed on macOS, too?

File: app/src/lib/dispatcher/app-shell.ts [hunk 2]
```diff
@@ -1,15 +1,56 @@
...
+
+function getEditorList(path: string ): IEditorInfo[] {
+  const result = new Array<IEditorInfo>()
+  result.push({
+    name: 'External Editor',
+  })
+  if (__DARWIN__) {
+    const osx = require('shell-osx')
+    if (osx.isAtomInstalled()) {
+    result.push({
```
Comment (j-f1) [inline app/src/lib/dispatcher/app-shell.ts]: This doesn’t appear to be indented.

File: app/src/lib/dispatcher/app-shell.ts [hunk 3]
```diff
@@ -1,15 +1,56 @@
...
+      name: 'Atom',
+    })
+    }
+  } else {
+    const win32 = require('./shell-win32')
+    if(win32.isVisualStudioInstalled() ) {
+    result.push({
```
Comment (j-f1) [inline app/src/lib/dispatcher/app-shell.ts]: This doesn’t look indented either.

File: app/src/lib/dispatcher/app-shell.ts (hunk 1 above)
Comment (shimpossible) [inline app/src/lib/dispatcher/app-shell.ts]: OSX part of the code is non-existent right now.  In fact it doesn't even check if Atom is installed
Plan to support VSCode.

Comment (j-f1) [inline app/src/lib/dispatcher/app-shell.ts]: If you’re creating your own local module, it should be imported with `require('./shell-osx')`.

File: app/npm-shrinkwrap.json [hunk 4]
```diff
@@ -1,1139 +1,928 @@
 {
   "name": "desktop",
   "version": "0.6.1",
+  "lockfileVersion": 1,
   "dependencies": {
     "7zip": {
-      "version": "0.0.6",
-      "from": "7zip@0.0.6",
-      "resolved": "https://registry.npmjs.org/7zip/-/7zip-0.0.6.tgz",
+      "version": "https://registry.npmjs.org/7zip/-/7zip-0.0.6.tgz",
```
Comment (shiftkey) [inline app/npm-shrinkwrap.json]: Looks like you're running `npm@5`, which we've not tested thoroughly. I fear changes like this `version` field are going to cause us issues down the track.

Comment (shimpossible) [inline app/npm-shrinkwrap.json]: thanks for catching that

**Comment by joshaber** [comment]
Ideally we'd allow the user to select whatever application they want, instead of hardcoding a supported list. Otherwise we're likely to get issues asking why we don't support _X_ editor 😄

I'm not very familiar with Windows—is there a reason to only support a fixed list?

**Comment by shimpossible** [comment]
Being able to edit the list would be nice.  I'm just trying to get a list as a first step.  Different editors work differently.  Vscode and atom can open a directory.  Visual studio and xcode require project files.  Changing to a dynamic list can be a later bunch of changes.

Not sure how I would change the settings ui to support a dynamic list, as I'm not sure what's needed yet.  I can add in some hooks for it and let someone else do the ui change.

**Comment by shimpossible** [comment]
😄 its configurable now.  should work on osx and win
Need to add a settings ui so the user can change things.   

Right now I do it in the console
```
localStorage.setItem('external-editors-.json', '[{"name":"Visual Studio", "cmd":"\\"C:\\\\Program Files (x86)\\\\Microsoft Visual Studio\\\\2017\\\\Community\\\\Common7\\\\IDE\\\\devenv.exe\\" \\"{path}\\""}]')
```

Might be better to store it by app instead of extension..  not sure yet

**Comment by saul** [comment]
Why not open with the user's default editor for a given extension?

**Comment by shimpossible** [comment]
It already opens with the default editor. Look at #2009
Theyre was also a request to open with multiple editors

File: app/src/lib/dispatcher/app-shell.ts [hunk 5]
```diff
@@ -18,8 +18,9 @@ export interface IAppShell {
   readonly openExternal: (path: string) => void
   readonly openItem: (path: string) => boolean
   readonly getEditors: (path: string) => Promise<IEditorLauncher[]>
-  readonly setEditors:  (ext: string, info: IEditorInfo) => void
+  readonly setEditors:  (ext: string, info: IEditorInfo[]) => void
```
Comment (j-f1) [inline app/src/lib/dispatcher/app-shell.ts]: `ReadonlyArray<IEditorInfo>`?

Comment (shimpossible) [inline app/src/lib/dispatcher/app-shell.ts]: Is that better than []?

Comment (j-f1) [inline app/src/lib/dispatcher/app-shell.ts]: It specifies that you can’t mutate the array, which is a good thing to verify if you don’t actually need to mutate it.

**Comment by shimpossible** [comment]
🎐

**Comment by rlabrecque** [comment]
Can you show some screenshots of what this looks like now in all the places it changed?

**Comment by shimpossible** [comment]
image

Right clicking on any file (if it has an association) shows the menu

...
open in external editor
open in <insert name here>

Right click on a repo also shows:
open in <insert name here>

**Comment by shiftkey** [comment]
@shimpossible apologies for not chiming in sooner, but I think we need to revisit the scope for this feature in 1.0 and keep it focused.

I wrote up some details in the original issue https://github.com/desktop/desktop/issues/2009#issuecomment-319113822 - could you have a look at that?

**Comment by shimpossible** [comment]
I agree.  This quickly got complicated.  And I needs a design that everyone agrees on.  But I had already started down the path and wanted to finish.  If this doesn't get used, or get refactored into something better that's fine

**Comment by BigChief45** [comment]
I remember some time ago using Github Desktop for Windows and it automatically could find Atom editor when it was installed in my machine. I could open the editor with the whole repository by clicking on the Atom logo shown on the top-right corner of Github Desktop.

However recently I downloaded and installed Github desktop and it looks completely different. Here is what it looked like before, sort of:

image

Can anyone tell me what happened? Was the whole UI changed? I could not find any mention of this in the release notes.

**Comment by joshaber** [comment]
>Can anyone tell me what happened?

Yes! The version you downloaded is the GitHub Desktop Beta. You can read more about it at https://github.com/blog/2362-announcing-git-integration-for-atom-and-github-desktop-beta.

If you want to download the Classic apps, scroll to the bottom of https://desktop.github.com for the links.

**Comment by BigChief45** [comment]
@joshaber 

Thanks for that. So if I understand correctly, Github Desktop Beta (the one I used before and the one that supported Atom Editor) had separate and different code bases for each platform. However a decision was made to re-write Github Desktop using Electron so that this new single code base (this repository I assume) could work accross all platforms.

**Comment by shimpossible** [comment]
There is a new version of desktop that is built in typescript. It's listed as beta.  If you scroll down to the bottom of https://desktop.GitHub.Com there is a link to the old one

**Comment by shiftkey** [comment]
@shimpossible thanks for your work on getting this going, but we ended up merging #2407 which will go out in the next update. Please feel free to open issues against it, so we can refine it down the track!

**Pull request closed without merging**